File: pages/refactorings/replace-mutable-fields-with-lenses/01-account-balance/After.scala
// Same, with a lens on the balance: the getter and the pure setter
// packed as one value.
object After:
  case class Account(balance: Int, bonus: Int)

  case class Lens[S, A](get: S => A, set: A => S => S):
    def modify(f: A => A): S => S = s => set(f(get(s)))(s)

  val balance: Lens[Account, Int] =
    Lens[Account, Int](_.balance, b => a => a.copy(balance = b))

  def withdraw(a: Account, amount: Int): Account =
    balance.modify(_ - amount)(a)

File: pages/refactorings/replace-mutable-fields-with-lenses/01-account-balance/After.hs
-- Same, with a lens on the balance: the getter and the pure setter
-- packed as one value.
module After where

data Account = Account { balance :: Int, bonus :: Int }
  deriving (Eq, Show)

data Lens s a = Lens { get :: s -> a, set :: a -> s -> s }

modify :: Lens s a -> (a -> a) -> s -> s
modify l f s = set l (f (get l s)) s

balanceLens :: Lens Account Int
balanceLens = Lens { get = balance, set = \b a -> a { balance = b } }

withdraw :: Account -> Int -> Account
withdraw a amount = modify balanceLens (subtract amount) a
